### DISCUSSION TITLE
Auth issue between Mobile app and VPN server

### DISCUSSION THREAD

**Discussion by gitanics** [strongswan/strongswan#2075] [Q&A]
Hello All,
Im facing issue with authentication when i try to connect using android mobile app to the vpn server, all details and config as below:

##############SERVER SIDE ##########
#########################################################
################### nano /etc/ipsec.conf ###################
config setup
    charondebug="ike 2, knl 2, cfg 2, net 2, esp 2, dmn 2, 100"
    uniqueids=no

conn %default
  ikelifetime=60m
  keylife=20m
  rekeymargin=3m
  keyingtries=1
  keyexchange=ikev2
  mobike=no
  ike=aes256-sha256-ecp256,aes256gcm16-prfsha256-ecp256,chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,3des-sha1-modp1024!
  esp=aes256-sha256,aes256-sha256-ecp256,aes256gcm16,aes256gcm16-ecp256,chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1,3des-sha1!
  leftcert=/etc/ssl/certs/myCA.crt

conn ipsec-ikv2-vpn
  left=%defaultroute
  auto=start
  rightsourceip=%dhcp
  rightdns=8.8.8.8,8.8.4.4
  rightsourceip=10.0.0.0/24
  right=%any
  rightid=%any
  rightauth=eap-dynamic
  eap_identity=%identity
  ikelifetime=60m
  keyexchange=ikev2
  keyingtries=1
  keylife=20m
  rekeymargin=3m
  dpddelay=0

###################  nano /etc/strongswan.conf ###################
charon {

    load_modular = yes

    plugins {
             eap_dynamic {
                 preferred = eap-mschapv2, eap-tls
              }
        include strongswan.d/charon/*.conf
    }
}

################### nano /etc/ipsec.secrets ####################
user1 : EAP "password1"
user2 : EAP "password2"

#############
root@pmm:/var/lib/strongswan# ipsec statusall
Status of IKE charon daemon (strongSwan 5.9.5, Linux 5.19.0-43-generic, x86_64):
uptime: 12 minutes, since Jan 22 10:30:07 2024
malloc: sbrk 2527232, mmap 0, used 1280624, free 1246608
worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 0
loaded plugins: charon aesni aes rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp agent xcbc hmac gcm drbg attr kernel-netlink resolve socket-default connmark stroke updown eap-mschapv2 xauth-generic counters
Virtual IP pools (size/online/offline):
10.0.0.0/24: 254/0/0
Listening IP addresses:
144.91.72.175
172.17.0.1
Connections:
ipsec-ikv2-vpn: %any...%any IKEv2
ipsec-ikv2-vpn: local: [C=AU, ST=Some-State, O=Internet Widgits Pty Ltd, CN=vpn-server] uses public key authentication
ipsec-ikv2-vpn: cert: "C=AU, ST=Some-State, O=Internet Widgits Pty Ltd, CN=vpn-server"
ipsec-ikv2-vpn: remote: uses EAP_DYNAMIC authentication with EAP identity '%any'
ipsec-ikv2-vpn: child: dynamic === dynamic TUNNEL
Security Associations (0 up, 0 connecting):
none

##############MOBILE SIDE ##########
Jan 22 09:30:12 00[DMN] +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
Jan 22 09:30:12 00[DMN] Starting IKE service (strongSwan 5.9.11, Android 13 - CPH2251_13.1.0.585(EX01)/2023-11-05, CPH2251 - OPPO/CPH2251/OPPO, Linux 4.19.191+, aarch64, org.strongswan.android)
Jan 22 09:30:12 00[LIB] providers loaded by OpenSSL: default legacy
Jan 22 09:30:12 00[LIB] loaded plugins: androidbridge charon android-log socket-default openssl nonce pkcs1 pem x509 xcbc kdf revocation eap-identity eap-mschapv2 eap-md5 eap-gtc eap-tls
Jan 22 09:30:12 00[JOB] spawning 16 worker threads
Jan 22 09:30:12 08[IKE] initiating IKE_SA android[8] to 144.91.72.175
Jan 22 09:30:12 08[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
Jan 22 09:30:12 08[NET] sending packet: from 10.230.78.163[36048] to 144.91.72.175[500] (948 bytes)
Jan 22 09:30:12 08[NET] received packet: from 144.91.72.175[500] to 10.230.78.163[36048] (305 bytes)
Jan 22 09:30:12 08[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
Jan 22 09:30:12 08[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_256
Jan 22 09:30:12 08[IKE] local host is behind NAT, sending keep alives
Jan 22 09:30:13 08[IKE] received cert request for "C=AU, ST=Some-State, O=Internet Widgits Pty Ltd, CN=vpn-server"
Jan 22 09:30:13 08[IKE] sending cert request for "C=AU, ST=Some-State, O=Internet Widgits Pty Ltd, CN=vpn-server"
Jan 22 09:30:13 08[IKE] establishing CHILD_SA android{8}
Jan 22 09:30:13 08[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) CERTREQ CPRQ(ADDR ADDR6 DNS DNS6) SA TSi TSr N(MOBIKE_SUP) N(NO_ADD_ADDR) N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
Jan 22 09:30:13 08[NET] sending packet: from 10.230.78.163[39538] to 144.91.72.175[4500] (448 bytes)
Jan 22 09:30:13 02[NET] received packet: from 144.91.72.175[4500] to 10.230.78.163[39538] (176 bytes)
Jan 22 09:30:13 02[ENC] parsed IKE_AUTH response 1 [ IDr EAP/FAIL ]
Jan 22 09:30:13 02[IKE] received EAP_FAILURE, EAP authentication failed
Jan 22 09:30:13 02[ENC] generating INFORMATIONAL request 2 [ N(AUTH_FAILED) ]
Jan 22 09:30:13 02[NET] sending packet: from 10.230.78.163[39538] to 144.91.72.175[4500] (80 bytes)

If someone faced the same issue, will be great to help me

Regards,
Mo

**Comment by tobiasbrunner** [comment]
You have to check the server log to see why the EAP authentication failed. Since no EAP exchange was initiated, a possibility is that the _eap-identity_ plugin is missing (if so, [see the FAQs](https://docs.strongswan.org/docs/5.9/support/faq.html#_plugin_is_missing)).

**Reply by gitanics** [reply]
Hi Tobiasbrunner,
Thank you for your reply.
As i can see it its loaded, check the  loaded plugins below :

root@pmm:~# ipsec statusall
Status of IKE charon daemon (strongSwan 5.9.5, Linux 5.19.0-43-generic, x86_64):
  uptime: 2 minutes, since Jan 24 13:11:11 2024
  malloc: sbrk 3428352, mmap 0, used 1463632, free 1964720
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 0
  **loaded plugins:** charon aesni aes rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp agent xcbc hmac gcm drbg attr kernel-netlink resolve socket-default connmark forecast farp stroke updown eap-identity eap-aka eap-md5 eap-gtc **eap-mschapv2** **eap-dynamic** eap-radius eap-tls eap-ttls eap-peap eap-tnc xauth-generic xauth-eap xauth-pam tnc-tnccs dhcp lookip error-notify certexpire led addrblock unity counters
Virtual IP pools (size/online/offline):
  10.0.0.0/24: 254/0/0
Listening IP addresses:
  144.91.72.175
  172.17.0.1
Connections:
ipsec-ikv2-vpn:  %any...%any  IKEv2
ipsec-ikv2-vpn:   local:  [C=AU, ST=Some-State, O=Internet Widgits Pty Ltd, CN=vpn-server] uses public key authentication
ipsec-ikv2-vpn:    cert:  "C=AU, ST=Some-State, O=Internet Widgits Pty Ltd, CN=vpn-server"
ipsec-ikv2-vpn:   remote: uses EAP_DYNAMIC authentication with EAP identity '%any'
ipsec-ikv2-vpn:   child:  dynamic === dynamic TUNNEL
Security Associations (0 up, 0 connecting):
  none

And when i check the available plugins:
root@pmm:~# ls /usr/lib/ipsec/plugins | grep eap
libstrongswan-eap-aka.so
libstrongswan-eap-dynamic.so
libstrongswan-eap-gtc.so
libstrongswan-eap-identity.so
libstrongswan-eap-md5.so
libstrongswan-eap-mschapv2.so
libstrongswan-eap-peap.so
libstrongswan-eap-radius.so
libstrongswan-eap-tls.so
libstrongswan-eap-tnc.so
libstrongswan-eap-ttls.so
libstrongswan-xauth-eap.so

Is that correct or i miss something?

**Reply by tobiasbrunner** [reply]
OK, again, check the log for why this fails.

**Reply by gitanics** [reply]
logs showing : 
Jan 24 13:56:39 pmm charon: 04[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_256
Jan 24 13:56:39 pmm charon: 04[CFG] received supported signature hash algorithms: sha256 sha384 sha512 identity
Jan 24 13:56:39 pmm charon: 04[IKE] remote host is behind NAT
Jan 24 13:56:39 pmm charon: 04[CFG] sending supported signature hash algorithms: sha256 sha384 sha512 identity
Jan 24 13:56:39 pmm charon: 04[ENC] generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
Jan 24 13:56:39 pmm charon: 04[NET] sending packet: from 144.91.72.175[500] to 85.255.237.119[25862] (280 bytes)
Jan 24 13:56:39 pmm charon: 03[NET] sending packet: from 144.91.72.175[500] to 85.255.237.119[25862]
Jan 24 13:56:39 pmm charon: 02[NET] received packet: from 85.255.237.119[62284] to 144.91.72.175[4500]
Jan 24 13:56:39 pmm charon: 02[NET] waiting for data on sockets
Jan 24 13:56:39 pmm charon: 05[NET] received packet: from 85.255.237.119[62284] to 144.91.72.175[4500] (432 bytes)
Jan 24 13:56:39 pmm charon: 05[ENC] parsed IKE_AUTH request 1 [ IDi N(INIT_CONTACT) CPRQ(ADDR ADDR6 DNS DNS6) SA TSi TSr N(MOBIKE_SUP) N(NO_ADD_ADDR) N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
Jan 24 13:56:39 pmm charon: 05[IKE] local endpoint changed from 144.91.72.175[500] to 144.91.72.175[4500]
Jan 24 13:56:39 pmm charon: 05[IKE] remote endpoint changed from 85.255.237.119[25862] to 85.255.237.119[62284]
Jan 24 13:56:39 pmm charon: 05[CFG] looking for peer configs matching 144.91.72.175[%any]...85.255.237.119[user1]
Jan 24 13:56:39 pmm charon: 05[CFG]   candidate "ipsec-ikv2-vpn", match: 1/1/28 (me/other/ike)
Jan 24 13:56:39 pmm charon: 05[CFG] selected peer config 'ipsec-ikv2-vpn'
Jan 24 13:56:39 pmm charon: 05[IKE] initiating EAP_IDENTITY method (id 0x00)
Jan 24 13:56:39 pmm charon: 05[IKE] processing INTERNAL_IP4_ADDRESS attribute
Jan 24 13:56:39 pmm charon: 05[IKE] processing INTERNAL_IP6_ADDRESS attribute
Jan 24 13:56:39 pmm charon: 05[IKE] processing INTERNAL_IP4_DNS attribute
Jan 24 13:56:39 pmm charon: 05[IKE] processing INTERNAL_IP6_DNS attribute
Jan 24 13:56:39 pmm charon: 05[IKE] peer supports MOBIKE
Jan 24 13:56:39 pmm charon: 05[IKE] no private key found for 'CN=server_domain_or_IP'
Jan 24 13:56:39 pmm charon: 05[ENC] generating IKE_AUTH response 1 [ N(AUTH_FAILED) ]
Jan 24 13:56:39 pmm charon: 05[NET] sending packet: from 144.91.72.175[4500] to 85.255.237.119[62284] (80 bytes)
Jan 24 13:56:39 pmm charon: 05[IKE] IKE_SA ipsec-ikv2-vpn[4] state change: CONNECTING => DESTROYING
Jan 24 13:56:39 pmm charon: 03[NET] sending packet: from 144.91.72.175[4500] to 85.255.237.119[62284]

i followed this way to generate the cert/ipsec config:
mkdir -p ~/pki/{cacerts,certs,private}
chmod 700 ~/pki
pki --gen --type rsa --size 4096 --outform pem > ~/pki/private/ca-key.pem

pki --self --ca --lifetime 3650 --in ~/pki/private/ca-key.pem --type rsa --dn "CN=VPN root CA" --outform pem > ~/pki/cacerts/ca-cert.pem

pki --gen --type rsa --size 4096 --outform pem > ~/pki/private/server-key.pem

pki --pub --in ~/pki/private/server-key.pem --type rsa \
    | pki --issue --lifetime 1825 \
        --cacert ~/pki/cacerts/ca-cert.pem \
        --cakey ~/pki/private/ca-key.pem \
        --dn "CN=144.91.72.175" --san server_domain_or_IP \
        --flag serverAuth --flag ikeIntermediate --outform pem \
    >  ~/pki/certs/server-cert.pem


sudo cp -r ~/pki/* /etc/ipsec.d/


Configurar StrongSwan

sudo mv /etc/ipsec.conf{,.original}

sudo nano /etc/ipsec.conf
config setup
    charondebug="ike 2, knl 2, cfg 2, net 2, esp 2, dmn 2, 100"
    uniqueids=no

conn ipsec-ikv2-vpn
    auto=add
    compress=no
    type=tunnel
    keyexchange=ikev2
    fragmentation=yes
    forceencaps=yes
    dpdaction=clear
    dpddelay=300s
    rekey=no
    left=%any
    leftid=144.91.72.175
    leftcert=server-cert.pem
    leftsendcert=always
    leftsubnet=0.0.0.0/0
    #right=%any
    #rightid=%any
    rightauth=eap-mschapv2
    rightsourceip=10.10.10.0/24
    rightdns=8.8.8.8,8.8.4.4
    rightsendcert=never
    eap_identity=%identity
    ike=aes256-sha256-ecp256,aes256gcm16-prfsha256-ecp256,chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,3des-sha1-modp102>
    esp=aes256-sha256,aes256-sha256-ecp256,aes256gcm16,aes256gcm16-ecp256,chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1,3des-sha1!

What im missing here?

**Reply by tobiasbrunner** [reply]
You need to load the private key that matches the server certificate in `ipsec.secrets` (i.e. add `: RSA server-key.pem`).

**Reply by gitanics** [reply]
Thank you for your help :)
Still same issue, now new config as below:

: RSA "server-key.pem"
user1 : EAP "password1"
user2 : EAP "password2"

do we need to add it as : RSA "server-key.pem" or : RSA server-key.pem , i tried both, same issue

the path of the cert and the keys as below:
/etc/ipsec.d/private/server-key.pem
/etc/ipsec.d/private/ca-key.pem

/etc/ipsec.d/certs/server-cert.pem

Are my cert and keys generation steps correct?

**Reply by tobiasbrunner** [reply]
Again, read the logs (the complete log will show you whether the key is loaded, or if there are any issues etc.)
